fast-danksharding/src/fast_danksharding.rs

```diff
@@ -255,12 +259,188 @@ pub fn main_flow() {
+    ////////////////////////////////
+    let br1_time = Instant::now();
+
+    // K0 = MSM_rows(C_rows) (256x1)
+    let mut K0 = commit_batch(&mut SRS, &mut C_rows, N_ROWS);
+    println!("K0 {:0.3?}", br1_time.elapsed());
+
+    // B0 = ECINTT_col(K0) N_POINTS x 1 (256x1)
+    let mut B0 = interpolate_points(&mut K0, &mut interpolate_column_domain);
+    println!("B0 {:0.3?}", br1_time.elapsed());
+
+    // K1 = ECNTT_col(MUL_col(B0, [1 u u^2 ...])) N_POINTS x 1 (256x1)
+    let K1 = evaluate_points_on_coset(&mut B0, &mut evaluate_column_domain, &mut column_coset);
+    println!("K1 {:0.3?}", br1_time.elapsed());
+
+    // K = [K0, K1]  // 2*N_POINTS x 1 (512x1 commitments)
+    let mut K: Vec<Point> = (0..2 * N_ROWS).map(|_| Point::zero()).collect();
+    K0.copy_to(&mut K[..N_ROWS]).unwrap();
+    K1.copy_to(&mut K[N_ROWS..]).unwrap();
+    println!("K {:0.3?}", br1_time.elapsed());
+    
+    println!("Branch1 {:0.3?}", br1_time.elapsed());
+    assert_eq!(K, get_debug_data_points_proj_xy1_vec("K.csv", 2 * N_ROWS));
+
+    ////////////////////////////////
+    println!("Branch 2");
+    ////////////////////////////////
+    let br2_time = Instant::now();
+
+    let mut D_rows = evaluate_scalars_on_coset_batch(&mut C_rows, &mut evaluate_row_domain, N_ROWS, &mut row_coset);
+    println!("D_both {:0.3?}", br2_time.elapsed());
+
+    let mut D_transposed = unsafe { DeviceBuffer::uninitialized(2 * N_ROWS * M_POINTS).unwrap() };
+    transpose_scalar_matrix(&mut D_transposed.as_device_ptr(), &mut D_in, M_POINTS, N_ROWS);
+    transpose_scalar_matrix(&mut D_transposed.as_device_ptr().wrapping_offset((N_ROWS * M_POINTS) as isize), &mut D_rows, M_POINTS, N_ROWS);
+
+    let mut C0 = interpolate_scalars_batch(&mut D_transposed, &mut interpolate_column_domain, 2 * M_POINTS);
+    let mut D_cols = evaluate_scalars_on_coset_batch(&mut C0, &mut evaluate_column_domain, 2 * M_POINTS, &mut column_coset);
+
+    let mut D = unsafe { DeviceBuffer::uninitialized(4 * N_ROWS * M_POINTS).unwrap() };
+    transpose_scalar_matrix(&mut D.as_device_ptr(), &mut D_transposed, N_ROWS, 2 * M_POINTS);
+    transpose_scalar_matrix(&mut D.as_device_ptr().wrapping_offset((2 * N_ROWS * M_POINTS) as isize), &mut D_cols, N_ROWS, 2 * M_POINTS);
+
+    let mut D_host_flat: Vec<ScalarField> = (0..4 * N_ROWS * FLOW_SIZE).map(|_| ScalarField::zero()).collect();
+    D.copy_to(&mut D_host_flat[..]).unwrap();
+    let D_host_flat = D_host_flat.into_iter().map(|x| { Scalar { s: x } }).collect::<Vec<_>>();
+    let D_host = D_host_flat.chunks(2 * M_POINTS).collect::<Vec<_>>();
+    assert_eq!(D_host, get_debug_data_scalars("D.csv", 2 * N_ROWS, 2 * M_POINTS));
+    
+    println!("Branch2 {:0.3?}", br2_time.elapsed());
+    let D_b4rbo = D_host.clone();
+
+    ////////////////////////////////
+    println!("Branch 3");
+    ////////////////////////////////
+    let br3_time = Instant::now();
+
+    //d0 = MUL_row(d[mu], [S]) 1x8192
+    let d0: Vec<_> = (0..2 * N_ROWS)
```

Comment:
actually "batch" version of this loop should be faster, pls look at https://github.com/ingonyama-zk/fast-danksharding/blob/0e80c7bd7c9e154023e93014e057a0eb10c561cc/fast-danksharding/src/fast_danksharding.rs#L262 -  and https://github.com/ingonyama-zk/fast-danksharding/blob/0e80c7bd7c9e154023e93014e057a0eb10c561cc/fast-danksharding/src/fast_danksharding.rs#L32 ondevice equivalent should be faster